version: '3.8'

services:
  web:
    image: cdaprod/cda-minio-control:latest
    environment:
      OPENAI_API_KEY: ${OPENAI_API_KEY}
      MINIO_ENDPOINT: minio:9000
      MINIO_ROOT_USER: ${MINIO_ROOT_USER}
      MINIO_ROOT_PASSWORD: ${MINIO_ROOT_PASSWORD}
    ports:
      - "8000:8000"
    networks:
      - app_network
    depends_on:
      - weaviate
      - minio
    deploy:
      replicas: 3
      update_config:
        parallelism: 2
        delay: 10s
      restart_policy:
        condition: on-failure

  minio:
    image: cdaprod/cda-minio:latest
    volumes:
      - minio_data:/data
    command: server /data --console-address ":9001"
    networks:
      - app_network
    secrets:
      - minio_root_user
      - minio_root_password

  weaviate:
    image: cdaprod/cda-weaviate:latest
    environment:
      OPENAI_API_KEY: ${OPENAI_API_KEY}
      AUTHENTICATION_ANONYMOUS_ACCESS_ENABLED: 'true'
      PERSISTENCE_DATA_PATH: '/var/lib/weaviate'
      QUERY_DEFAULTS_LIMIT: 25
      DEFAULT_VECTORIZER_MODULE: 'text2vec-openai'
      ENABLE_MODULES: 'backup-s3, text2vec-openai'
      BACKUP_S3_BUCKET: 'weaviate-backups'
      BACKUP_S3_ENDPOINT: 'minio:9000'
      BACKUP_S3_ACCESS_KEY_ID: ${MINIO_ROOT_USER}
      BACKUP_S3_SECRET_ACCESS_KEY: ${MINIO_ROOT_PASSWORD}
      BACKUP_S3_USE_SSL: 'false'
      CLUSTER_HOSTNAME: 'node1'
    volumes:
      - weaviate_data:/var/lib/weaviate
    networks:
      - app_network
    depends_on:
      - minio

  nginx:
    image: cdaprod/cda-nginx:latest
    ports:
      - "80:80"
      - "443:443"
    volumes:
      - ./nginx/nginx.conf:/etc/nginx/nginx.conf
    networks:
      - app_network

  jupyterlab:
    build: 
      context: ./jupyter
    image: cdaprod/cda-jupyterlab:latest
    ports:
      - "8888:8888"
    volumes:
      - ../usb/001/002:/dev/bus/usb/001/002
      - jupyter_data:/home/jovyan/work
    networks:
      - app_network
    environment:
      - JUPYTER_ENABLE_LAB=yes
    privileged: true
    deploy:
      replicas: 1
      restart_policy:
        condition: on-failure
    command: ["jupyter", "lab", "--ip=0.0.0.0", "--allow-root", "--NotebookApp.token=''", "--NotebookApp.password=''"]

  tailscale:
    image: tailscale/tailscale
    container_name: tailscale
    cap_add:
      - NET_ADMIN
      - SYS_MODULE
    security_opt:
      - apparmor:unconfined
    volumes:
      - /dev/net/tun:/dev/net/tun
      - tailscale_data:/var/lib/tailscale
    environment:
      - TS_AUTH_KEY=${TS_AUTH_KEY}
    networks:
      - app_network
    command: tailscaled

networks:
  app_network:
    driver: overlay

volumes:
  tailscale_data:
    driver: local
  minio_data:
  weaviate_data:
  jupyter_data:
    driver: local
    driver_opts:
      type: none
      device: /opt/jupyter_data
      o: bind

secrets:
  minio_root_user:
    external: true
  minio_root_password:
    external: true